vault-agent/user_data.yaml.tpl

```diff
@@ -0,0 +1,131 @@
+#cloud-config
+merge_how:
+ - name: list
+   settings: [append, no_replace]
+ - name: dict
+   settings: [no_replace, recurse_list]
+
+%{ if install_dependencies ~}
+users:
+  - name: vault
+    system: true
+    lock_passwd: true
+%{ endif ~}
+
+write_files:
+  # Vault Agent Role ID
+  - path: /etc/vault-agent.d/role-id
+    owner: root:root
+    permissions: "0644"
+    content: "${vault_agent.auth_method.config.role_id}"
+
+  # Vault Agent Secret ID
+  - path: /etc/vault-agent.d/secret-id
+    owner: root:root
+    permissions: "0644"
+    content: "${vault_agent.auth_method.config.secret_id}"
+
+%{ if vault_agent.vault_ca_cert != "" ~}
+  # Vault Agent CA Certificate
+  - path: /etc/vault-agent.d/tls/ca.crt
+    owner: root:root
+    permissions: "0440"
+    content: |
+      ${vault_agent.vault_ca_cert}
+%{ endif ~}
+
+  # Vault Agent configuration
+  - path: /etc/vault-agent.d/agent.hcl
+    owner: root:root
+    permissions: "0644"
+    content: |
+      exit_after_auth = false
+      pid_file = "/var/run/vault-agent.pid"
+
+      auto_auth {
+        method "${vault_agent.auth_method.type}" {
+          config = {
+            role_id_file_path = "/etc/vault-agent.d/role-id"
+            secret_id_file_path = "/etc/vault-agent.d/secret-id"
+            remove_secret_id_file_after_reading = false
+          }
+        }
+
+        sink "file" {
+          config = {
+            path = "/etc/vault-agent.d/agent-token"
+          }
+        }
+      }
+
+      vault {
+        address = "${vault_agent.vault_address}"
+%{ if vault_agent.vault_ca_cert != "" ~}
+        ca_cert = "/etc/vault-agent.d/tls/ca.crt"
+%{ endif ~}
+      }
+
+      auto_reload {
+        enabled = true
+      }
+
+%{ for template in vault_agent.templates ~}
+      template {
+        source      = "${template.source_path}"
+        destination = "${template.destination_path}"
+        command     = "systemctl restart ${template.service_name}"
+      }
+%{ endfor ~}
+
+%{ if vault_agent.agent_config != "" ~}
+      ${vault_agent.agent_config}
+%{ endif ~}
```

Comment:
I think I'd rename this something like "extra_config" to make it clear it is optional extra configuration the call of the module can pass.